File: Day10/bin/tokenizer.cpp
#include "tokenizer.h"

#include <regex>
#include <map>
#include <algorithm>

namespace tokenizer
{
	std::vector<TokenDef> definitions;

	void tokenize(const std::string& src, std::vector<Token>& tokens)
	{
		std::vector<TokenMatch> matches;
		for (int i = 0; i < definitions.size(); i++)
		{
			findMatches(src, definitions[i], matches);
		}

		std::map<int, std::vector<TokenMatch>> groupedMatches;
		for (const TokenMatch& m : matches)
		{
			groupedMatches[m.start].push_back(m);
		}

		const TokenMatch* lastMatch = nullptr;
		for (std::pair<int, std::vector<TokenMatch>> e : groupedMatches)
		{
			std::sort(groupedMatches[e.first].begin(), groupedMatches[e.first].end(), [](const TokenMatch& a, const TokenMatch& b) {
				return a.precedence > b.precedence;
			});
			const TokenMatch* bestMatch = &groupedMatches[e.first].front();
			if (lastMatch != nullptr && bestMatch->start < lastMatch->end)
			{
				continue;
			}

			Token token;
			token.type = bestMatch->type;
			token.val = bestMatch->val;
			tokens.push_back(token);

			lastMatch = bestMatch;
		}
	}

	void findMatches(const std::string& input, const TokenDef& def, std::vector<TokenMatch>& matches)
	{
		std::cregex_iterator empty;
		for (auto it = std::cregex_iterator(input.c_str(), input.c_str() + input.length(), def.regex); it != empty; it++)
		{
			TokenMatch match;
			match.start = it->position(0);
			match.end = it->position(0) + it->length(0);
			match.val = it->str(0);
			match.type = def.type;
			match.precedence = def.precedence;
			matches.push_back(match);
		}
	}

	void addDefinition(int type, std::string regex, int precedence)
	{
		TokenDef def;
		def.type = type;
		def.regex = std::regex(regex);
		def.precedence = precedence;
		definitions.push_back(def);
	}

	void removeDefinitions()
	{
		definitions.clear();
	}
}

File: compiler/src/rb/compiler/Tokenizer.java
package rb.compiler;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.Vector;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Tokenizer {
	static Vector<TokenDef> definitions;

	static {
		definitions = new Vector<TokenDef>();
	}

	public static void run(String src, List<Token> tokens) {
		List<TokenMatch> matches = new ArrayList<TokenMatch>();
		for (TokenDef def : definitions) {
			findMatches(src, def, matches);
		}

		Map<Integer, List<TokenMatch>> groupedMatches = new TreeMap<Integer, List<TokenMatch>>();
		for (TokenMatch match : matches) {
			if (groupedMatches.get(match.start) == null) {
				groupedMatches.put(match.start, new ArrayList<TokenMatch>());
			}
			groupedMatches.get(match.start).add(match);
		}

		TokenMatch lastMatch = null;
		for (int i : groupedMatches.keySet()) {
			Collections.sort(groupedMatches.get(i), new Comparator<TokenMatch>() {
				@Override
				public int compare(TokenMatch o1, TokenMatch o2) {
					return o1.precedence > o2.precedence ? 1 : -1;
				}
			});
			TokenMatch bestMatch = groupedMatches.get(i).get(0);
			if (lastMatch != null && bestMatch.start < lastMatch.end) {
				continue;
			}

			Token token = new Token();
			token.type = bestMatch.type;
			token.val = bestMatch.val;
			tokens.add(token);

			lastMatch = bestMatch;
		}
	}

	private static void findMatches(String input, TokenDef def, List<TokenMatch> matches) {
		Pattern pattern = Pattern.compile(def.regex);
		Matcher matcher = pattern.matcher(input);
		while (matcher.find()) {
			TokenMatch match = new TokenMatch();
			match.start = matcher.start();
			match.end = matcher.end();
			match.val = input.substring(match.start, match.end);
			match.type = def.type;
			match.precedence = def.precedence;
			matches.add(match);
		}
	}

	public static void addDefinition(int type, String regex, int precedence) {
		TokenDef def = new TokenDef();
		def.type = type;
		def.regex = regex;
		def.precedence = precedence;
		definitions.add(def);
	}

	public static void removeDefinitions() {
		definitions.clear();
	}

	public static class Token {
		public int type;
		public String val;
	}

	public static class TokenDef {
		public String regex;
		public int type;
		public int precedence;
	}

	public static class TokenMatch {
		public int type;
		public String val;
		public int start, end;
		public int precedence;
	}
}
